### PR TITLE
[ISSUE-22] also mark classes that are only repackaged from javax => jakarta as transformed

### PR DESCRIPTION
**Pull request by scottmarlow**

### CODE REVIEW

File: impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java [hunk 1]
```diff
@@ -423,8 +421,14 @@ public void clearTransformationState() {
 
     @Override
     public Resource transform(final Resource r) {
+        
+        clearTransformationState(); // clear transformation state for each resource, prior to transformation
+        
         String oldResourceName = r.getName();
         String newResourceName = replacePackageName(oldResourceName, false);
+        if (!newResourceName.equals(oldResourceName)) {  // any file rename counts as a transformation 
+            setClassTransformed(true);
```
Comment (ropalka) [inline impl/asm-based-transformer/src/main/java/org/wildfly/transformer/asm/TransformerImpl.java:430]: This is not sufficient as constant pool isn't updated properly.
If you repackage the class only the classloading process will fail.
Here's simple reproducer:
---
public class E {
    public static void main(String[] args) throws Exception {
	final String s1 = "jakarta.servlet.SingleThreadModel";
	Class c = Class.forName(s1);
	System.out.println(c);
    }
}
---
Exception thrown:
---
[blackhole][/home/opalka]>java -cp jakarta-servlet-api-2.3-asm-transformer.jar:. E
Exception in thread "main" java.lang.NoClassDefFoundError: jakarta/servlet/SingleThreadModel (wrong name: javax/servlet/SingleThreadModel)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at E.main(E.java:5)
---
---

**Comment by scottmarlow** [comment]
For EE 8 => EE 9 transformation, what is the use case for repackaging classes?  Is it only for EE SPEC API jars?  

There is https://asm.ow2.io/javadoc/org/objectweb/asm/commons/ClassRemapper.html that I will look at.

**Comment by scottmarlow** [comment]
I might have a solution, need to add a ModuleVisitor to rename packages I think.

**Comment by ropalka** [comment]
Yes, it is for repackaging EE8 API jars to EE9 @scottmarlow

**Comment by ropalka** [comment]
Module visitor is targeting only Java 9+ isn't it?

**Comment by scottmarlow** [comment]
It may well be, I stepped into ASM and it never used that code, so it must be JPMS modules only.  Instead, I used the ClassVistor#visit(final int version,final int access,final String name,final String signature,final String superName,final String[] interfaces) callback and that seemed to work.

It is hard to tell for sure, due to ASM not removing old javax entries from constant pool. 

Output from java -cp jboss-servlet-api_4.0_spec-2.0.0.Final.jar:. E
interface jakarta.servlet.SingleThreadModel:
interface jakarta.servlet.SingleThreadModel

**Comment by ropalka** [comment]
I can say for sure it is doing required transformation. Here's output from javap -v SingleThreadModel.class

---

Classfile /home/opalka/SingleThreadModel.class
  Last modified Apr 24, 2020; size 168 bytes
  MD5 checksum cbd7a7dfaab0a37fdba07ca159078612
  Compiled from "SingleThreadModel.java"
public interface jakarta.servlet.SingleThreadModel
  minor version: 3
  major version: 45
  flags: ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT
Constant pool:
  #1 = Class              #5              // javax/servlet/SingleThreadModel
  #2 = Class              #6              // java/lang/Object
  #3 = Utf8               SourceFile
  #4 = Utf8               SingleThreadModel.java
  #5 = Utf8               javax/servlet/SingleThreadModel
  #6 = Utf8               java/lang/Object
  #7 = Utf8               jakarta/servlet/SingleThreadModel
  #8 = Class              #7              // jakarta/servlet/SingleThreadModel
{
}
SourceFile: "SingleThreadModel.java"

---

**Comment by ropalka** [comment]
Thanks @scottmarlow

**Pull request merged**